rust/src/mime/mime.rs

```diff
@@ -0,0 +1,605 @@
+    if let Some(value) = mime_find_header_token(hbuf, tbuf, &mut sections_values) {
+        // limit the copy to the supplied buffer size
+        if value.len() <= RS_MIME_MAX_TOKEN_LEN {
+            outbuf[..value.len()].clone_from_slice(value);
+        } else {
+            outbuf.clone_from_slice(&value[..RS_MIME_MAX_TOKEN_LEN]);
+        }
+        *outlen = value.len() as u32;
+        return true;
+    }
+    return false;
+}
+
+#[derive(Debug)]
+enum MimeParserState {
+    Start = 0,
+    Header = 1,
+    HeaderEnd = 2,
+    Chunk = 3,
+    BoundaryWaitingForEol = 4,
+}
+
+impl Default for MimeParserState {
+    fn default() -> Self {
+        MimeParserState::Start
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct MimeStateHTTP {
+    boundary: Vec<u8>,
+    filename: Vec<u8>,
+    state: MimeParserState,
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialOrd, PartialEq, Eq)]
+pub enum MimeParserResult {
+    MimeNeedsMore = 0,
+    MimeFileOpen = 1,
+    MimeFileChunk = 2,
+    MimeFileClose = 3,
+}
+
+fn mime_parse_skip_line(input: &[u8]) -> IResult<&[u8], MimeParserState> {
+    let (input, _) = take_till(|ch: u8| ch == b'\n')(input)?;
+    let (input, _) = char('\n')(input)?;
+    return Ok((input, MimeParserState::Start));
+}
+
+fn mime_parse_boundary_regular<'a>(
+    boundary: &[u8], input: &'a [u8],
+) -> IResult<&'a [u8], MimeParserState> {
+    let (input, _) = tag(boundary)(input)?;
+    let (input, _) = take_till(|ch: u8| ch == b'\n')(input)?;
+    let (input, _) = char('\n')(input)?;
+    return Ok((input, MimeParserState::Header));
+}
+
+// Number of characters after boundary, without end of line, before changing state to streaming
+const MIME_BOUNDARY_MAX_BEFORE_EOL: usize = 128;
+const MIME_HEADER_MAX_LINE: usize = 4096;
+
+fn mime_parse_boundary_missing_eol<'a>(
+    boundary: &[u8], input: &'a [u8],
+) -> IResult<&'a [u8], MimeParserState> {
+    let (input, _) = tag(boundary)(input)?;
+    let (input, _) = take(MIME_BOUNDARY_MAX_BEFORE_EOL)(input)?;
+    return Ok((input, MimeParserState::BoundaryWaitingForEol));
+}
+
+fn mime_parse_boundary<'a>(boundary: &[u8], input: &'a [u8]) -> IResult<&'a [u8], MimeParserState> {
+    let r = mime_parse_boundary_regular(boundary, input);
+    if r.is_ok() {
+        return r;
+    }
+    let r2 = mime_parse_skip_line(input);
+    if r2.is_ok() {
+        return r2;
+    }
+    return mime_parse_boundary_missing_eol(boundary, input);
+}
+
+fn mime_consume_until_eol(input: &[u8]) -> IResult<&[u8], bool> {
+    return alt((value(true, mime_parse_skip_line), value(false, rest)))(input);
+}
+
+pub fn mime_parse_header_line(input: &[u8]) -> IResult<&[u8], &[u8]> {
+    let (input, name) = take_till(|ch: u8| ch == b':')(input)?;
+    let (input, _) = char(':')(input)?;
+    let (input, _) = take_while(is_mime_space)(input)?;
+    return Ok((input, name));
+}
+
+// s2 is already lower case
+pub fn rs_equals_lowercase(s1: &[u8], s2: &[u8]) -> bool {
```

Comment:
This isn't exposed to C, can drop the `rs` as its namespaced.

Reply:
ok, renamed as `slice_equals_lowercase`